include: kernel: Add macro to avoid stack overhead

albertofloyd opened the pull request:
When CONFIG_USER_SPACE is enabled threads end up allocating space for a privileged stack.
When a thread will run in kernel mode, this extra space is not needed.

k_thread_create() allows to make this distinction passing an object defined using K_KERNEL_STACK_DEFINE macro.
https://docs.zephyrproject.org/3.1.0/kernel/services/threads/index.html#kernel-only-stacks

K_THREAD_DEFINE macro does not allow to make the same distinction. 
Need an additional macro to be able to define threads intended to run always in kernel mode.

albertofloyd committed:
include: kernel: Allow to define threads without a privileged stack

When CONFIG_USER_SPACE is enabled threads end up allocating
space for a privileged stack.
When a thread will run in kernel mode, this extra space is not
needed.

k_thread_create() allows to make this distinction passing an object
defined using K_KERNEL_STACK_DEFINE macro.

K_THREAD_DEFINE macro does not allow to make the same distinction,
Need an additional macro to be able to define threads intended
to run always in kernel mode.

Signed-off-by: Jose Alberto Meza <jose.a.meza.arellano@intel.com>

andyross reviewed (changes requested):
So, first, new public APIs need docs so readers understand what the purpose of an API is.  Which brings me to...

I don't quite understand the purpose of this API.  A thread with a priviledged stack is a user mode thread, one without is a kernel thread.  It seems like you're just creating a new kind of kernel thread and naming it "supervisory thread"?

Are you just trying to have a THREAD_DEFINE() macro for static kernel threads?  If so, why not name it K_KERNEL_THREAD_DEFINE(), and adjust the scheme in the headers such that it shares the same implementation with a non-userspace K_THREAD_DEFINE()?

ceolin commented:
> So, first, new public APIs need docs so readers understand what the purpose of an API is. Which brings me to...
> 
> I don't quite understand the purpose of this API. A thread with a priviledged stack is a user mode thread, one without is a kernel thread. It seems like you're just creating a new kind of kernel thread and naming it "supervisory thread"?
> 
> Are you just trying to have a THREAD_DEFINE() macro for static kernel threads? If so, why not name it K_KERNEL_THREAD_DEFINE(), and adjust the scheme in the headers such that it shares the same implementation with a non-userspace K_THREAD_DEFINE()?

I think the problem trying to be addressed here is that we always end up allocating space for a privileged stack in all threads when userspace is enabled. If you know that a thread will always run in kernel mode this extra space is not needed. Sounds reasonable to me this, but there are things missing, we need to put this info into the thread so if one tries to move it to usermode it properly fails. Right now you are just avoiding the additional "privileged stack" memory allocation but what happens if we try to make it an usermode thread ?

The names are really confusing, regarding sharing code it will need to change those macros if I remember correctly.

albertofloyd commented:
> So, first, new public APIs need docs so readers understand what the purpose of an API is. Which brings me to...
> 
> I don't quite understand the purpose of this API. A thread with a priviledged stack is a user mode thread, one without is a kernel thread. It seems like you're just creating a new kind of kernel thread and naming it "supervisory thread"?
> 
> Are you just trying to have a THREAD_DEFINE() macro for static kernel threads? If so, why not name it K_KERNEL_THREAD_DEFINE(), and adjust the scheme in the headers such that it shares the same implementation with a non-userspace K_THREAD_DEFINE()?

Updated the PR comment with a clearer intention (i hope), added documentation to use KERNEL_THREAD to keep it inline with stack naming convention in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/kernel.h#L285

K_THREAD_STACK_DEFINE() - For stacks that may support either user or
 *   supervisor threads.
 * - K_KERNEL_STACK_DEFINE() - For stacks that may support supervisor
 *   threads only. These stacks use less memory if CONFIG_USERSPACE is
 *   enabled.

albertofloyd commented:
> > So, first, new public APIs need docs so readers understand what the purpose of an API is. Which brings me to...
> > I don't quite understand the purpose of this API. A thread with a priviledged stack is a user mode thread, one without is a kernel thread. It seems like you're just creating a new kind of kernel thread and naming it "supervisory thread"?
> > Are you just trying to have a THREAD_DEFINE() macro for static kernel threads? If so, why not name it K_KERNEL_THREAD_DEFINE(), and adjust the scheme in the headers such that it shares the same implementation with a non-userspace K_THREAD_DEFINE()?
> 
> I think the problem trying to be addressed here is that we always end up allocating space for a privileged stack in all threads when userspace is enabled. If you know that a thread will always run in kernel mode this extra space is not needed. Sounds reasonable to me this, but there are things missing, we need to put this info into the thread so if one tries to move it to usermode it properly fails. 

@ceolin  Isn't passing a kernel-only stack enough to guarantee that?  https://docs.zephyrproject.org/3.1.0/kernel/services/threads/index.html#kernel-only-stacks

> Right now you are just avoiding the additional "privileged stack" memory allocation but what happens if we try to make it an usermode thread ?
> 
> The names are really confusing, regarding sharing code it will need to change those macros if I remember correctly.

@ceolin could you elaborate about the kind of conditions to check. I'm not familiar with this Zephyr mechanism, can threads defined this way be updated dynamically to use a different stack or do you mean compilation time?

andyross commented on include/zephyr/kernel.h:
```diff
@@ -725,6 +725,48 @@ struct _static_thread_data {
+ *
+ * @internal It has been observed that the x86 compiler by default aligns
+ * these _static_thread_data structures to 32-byte boundaries, thereby
+ * wasting space. To work around this, force a 4-byte alignment.
```
Not understanding the note here?  There's no 4-byte alignment done here.  Also 4 bytes is fine for i386 but definitely wrong in general.  The sysv ABI on x86_64 requires a 16 byte alignment, surely other architectures have rules of their own.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -725,6 +725,48 @@ struct _static_thread_data {
+ */
+#define K_KERNEL_THREAD_DEFINE(name, stack_size,                                \
+			entry, p1, p2, p3,                               \
+			prio, options, delay)                            \
```
FWIW: the implementation here is (or should be, anyway) identical to the way K_THREAD_DEFINE() works when CONFIG_USERSPACE=n.  It would be good to unify the two to save the cut/paste bandwidth and reduce bug surface.

andyross reviewed (commented):
Some nitpicks, but in general I'm fine with the API.

dcpleung commented:
You will need some mechanisms to prevent such a thread from entering user mode via `k_thread_user_mode_enter()`.

andyross commented:
>  You will need some mechanisms to prevent such a thread from entering user mode via k_thread_user_mode_enter().

Actually I think that's OK?  I mean, it won't work and will surely panic the system, but it's a call that only works from kernel mode anyway.   It's no different than a thread created without the USER flag via k_thread_create().

dcpleung commented:
> Actually I think that's OK? I mean, it won't work and will surely panic the system, but it's a call that only works from kernel mode anyway. It's no different than a thread created without the USER flag via k_thread_create().

`k_thread_user_mode_enter()` adds `K_USER` to the thread option and will make the current running kernel thread into a user thread. For example, `z_main_stack` is "user capable" so you can make the main thread (with K_USER) into a user thread without any issue. The `z_stack_is_user_capable()` is going to return true because the stack object appears in the kobject list (even for stacks used for kernel threads). So without all the extra space needed for user thread stacks, it is possible different threads under a domain can corrupt each others' stacks. I think it is better to have some extra checks to deny user mode entry if it is a strict kernel-only thread.

dcpleung commented:
I think the original idea is that you can have a kernel thread running all the necessary code to grant access to various kernel objects, and then turns itself into user thread. So there is no need to have separate threads granting access, and thus no need to manually serialize execution. Hence the "backend" does not really distinguish between kernel and user threads. This will need some time to unravel.
